cpp/src/arrow/array/builder_base.h

```diff
@@ -123,6 +123,14 @@ class ARROW_EXPORT ArrayBuilder {
   Status AppendScalar(const Scalar& scalar, int64_t n_repeats);
   Status AppendScalars(const ScalarVector& scalars);
 
+  /// \brief Append a range of values from an array.
+  ///
+  /// The given array must be the same type as the builder.
+  virtual Status AppendArraySliceUnchecked(const ArrayData& array, int64_t offset,
```

Comment:
We can drop "Unchecked" from the name, I think, as long as the contract is documented.